otherlibs/stdune/src/compact_position.ml

```diff
@@ -0,0 +1,118 @@
+module Position = struct
+  type t = int
+
+  let size = 21
+
+  let mask_bol = 0x1FFFFF (* 21 bits set *)
+
+  let mask_lnum = mask_bol lsl size
+
+  let mask_cnum = mask_lnum lsl size
+
+  let small_enough =
+    let max_size = 1 lsl size in
+    let test int = int <= max_size in
+    fun { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname = _ } ->
+      test pos_bol && test pos_cnum && test pos_lnum
+
+  let of_position { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname = _ } =
+    let t = pos_cnum land mask_bol in
+    let t = (t lsl size) lor (pos_lnum land mask_bol) in
+    (t lsl size) lor (pos_bol land mask_bol)
+
+  let bol t = t land mask_bol
+
+  let lnum t = (t land mask_lnum) lsr size
+
+  let cnum t = (t land mask_cnum) lsr (size * 2)
+
+  let to_position t ~fname:pos_fname =
+    let pos_bol = bol t in
+    let pos_cnum = cnum t in
+    let pos_lnum = lnum t in
+    { Lexing.pos_bol; pos_cnum; pos_lnum; pos_fname }
+end
+
+module Same_line_loc = struct
+  type t = int
+
+  let size = 15
+
+  let mask_bol = 0x7FFF (* 15 bits set *)
+
+  let mask_lnum = mask_bol lsl size
+
+  let mask_start_cnum = mask_lnum lsl size
+
+  let mask_stop_cnum = mask_start_cnum lsl size
+
+  let create ~bol ~lnum ~start_cnum ~stop_cnum =
+    let t = stop_cnum land mask_bol in
+    let t = (t lsl size) lor (start_cnum land mask_bol) in
+    let t = (t lsl size) lor (lnum land mask_bol) in
+    (t lsl size) lor (bol land mask_bol)
+
+  let bol t = t land mask_bol
+
+  let lnum t = (t land mask_lnum) lsr size
+
+  let start_cnum t = (t land mask_start_cnum) lsr (size * 2)
+
+  let stop_cnum t = (t land mask_stop_cnum) lsr (size * 3)
+
+  let small_enough =
+    let max_size = 1 lsl size in
+    fun int -> int <= max_size
+
+  let to_loc t ~fname:pos_fname =
+    let pos_lnum = lnum t in
+    let pos_bol = bol t in
+    let start =
+      { Lexing.pos_fname; pos_lnum; pos_bol; pos_cnum = start_cnum t }
+    in
+    let stop = { start with pos_cnum = stop_cnum t } in
+    { Lexbuf.Loc.start; stop }
+
+  let start t ~fname:pos_fname =
+    let pos_lnum = lnum t in
+    let pos_bol = bol t in
+    { Lexing.pos_fname; pos_lnum; pos_bol; pos_cnum = start_cnum t }
+
+  let stop t ~fname:pos_fname =
+    let pos_lnum = lnum t in
+    let pos_bol = bol t in
+    { Lexing.pos_fname; pos_lnum; pos_bol; pos_cnum = stop_cnum t }
+end
+
+include Position
+
+type of_loc =
+  | Same_line of Same_line_loc.t
+  | Loc of
+      { start : t
+      ; stop : t
+      }
+  | Location_does_not_fit
```

Comment:
> Of course that makes the memory usage worse in that sitaution as another cell is allocated but this would be hopefully a rare case.

That's precisely why I omitted it. This Compact_position is a low level, private API, so we can keep it a little ugly for performance's sake.